man/im_stats.3
.TH IM_STATS 3 "24 October 1992"
.SH NAME
im_stats \- calculate many image statistics in a single pass
.SH SYNOPSIS
.B #include <vips/vips.h>

.B DOUBLEMASK *im_stats(in)
.br
.B IMAGE *in;
.SH DESCRIPTION
Find many image statistics in a single pass through the PELs. Returns a
DOUBLEMASK of 6 columns by n+1 (where n is number of bands in image in) rows.
Columns are statistics, and are, in order: minimum, maximum, sum, sum of
squares, mean, standard deviation. Row 0 has statistics for all bands
together, row 1 has stats for band 1, and so on.

Non-complex images only! 
.SH RETURN VALUE
NULL on error.
.SH SEE\ ALSO
im_avg(3), im_deviate(3)
.SH COPYRIGHT
.br
National Gallery, 1992


NEWS
VIPS changed from 7.10 to 7.12 (not exhaustive, see ChangeLog for details)

Non-backwards compatible changes
================================

- IMAGE->Hist is deprecated, use im_history_get() instead
- im_append_Hist() removed
- IMAGE->Bbits is deprecated (now ignored)
- im_region_local() replaced by im_region_buffer()

VIPS enhancements
=================

- new backwards and forwards compatible VIPS file format: it has a new 
  metadata system that efficiently shares and propogates ICC profiles,
  EXIF data, etc. & whatever
- speed ups: new pixel buffer cache saves and reuses computations, uses liboil 
  where possible for a small speedup
- SMP scaling: simplified lock system improves SMP scaling, double-buffered 
  image writes overlap computation 
- lower mem use: new mmap window cache shares file mappings, history buffers 
  share and reuse image history metadata
- built-in support for new image formats: OpenEXR, Analyze7, CSV
- RGB16 and GREY16 image hints
- uses GOption for much better command-line argument handling
- better C++ binding: namespaces, VError derives from std::exception, image
  and number vector constants
- python binding 
- gcc sentinel attributes added
- added GValue as an arg type
- added im_benchmark() and friends for testing
- new functions: 
	im_analyze2vips(), im_benchmark*(), im_buildlut(),
	im_concurrency_get/set(), im_contrast_surface(),
	im_contrast_surface_raw(), im_copy_from(), im_copy_morph(),
	im_cp_desc_array(), im_cp_descv(), im_extract_areabands(),
	im_flood_blob_copy(), im_get_option_group(), im_grid(),
	im_header_exists(), im_header_map(), im_history_get(),
	im_invalidate(), im_isscalar(), im_lineset(), im_linreg(),
	im_meta*(), im_msb(), im_norm_dmask(), im_project(),
	im_raw2vips(), IM_REGION_ADDR_TOPLEFT(), im_tile_cache(),
	im_video_test()


VIPS changed from 7.8 to 7.10 (not exhaustive, see ChangeLog for details)

Non-backwards compatible changes
================================

- no longer generates vips-config script, instead it makes .pc files for
  pkg-config ... replace any "vips-config" with "pkg-config vips-7.10"
- origin hint semantics changed ... it now records the position in the output
  image of the origin of the input image
- all float to int casts on pixel values are now floor() rather than rint(),
  for consistency with C casting semantics

VIPS enhancements
=================

- nip reworked for gtk+-2.4, now called nip2
- now based on glib, so VIPS no longer has it's own thread, plugin and data
  structure libraries
- i18n support, although not quite complete yet
- new error message API to make i18n possible (compatibility macros mean the
  old API still works)
- new 'start up VIPS' call, im_init_world(), sets up i18n and threads
  (optional to call it, but i18n won't fully work unless you do)
- C programs now just need "#include <vips/vips.h>", no other includes
  required
- wraps the Pango font rendering library, see im_text()
- new function flag: NOCACHE ... means the result of the call should not be
  cached, useful for video grab functions and paintbox functions
- im_embed() now supports a wider range of embedding styles, including a fast
  edge stretch
- all area operations use the new embed style to expand their input rather 
  than their output, so there are no more black borders on filtered images
- other new functions: im_render(), im_cache(), im_extract_bands(), 
  im_copy_swap(), im_rint(), im_make_xy(), im_init_world(), im_error(),
  im_warn(), im_diag(), im_rank_image(), im_replicate()
- many fixes and improvements to old funcs
- configure and portability improvements

Bug fixes
=========

- all serious bug fixes got back-ported to 7.8, I think


VIPS changes from 7.6 to 7.8 (again, not exhaustive, see ChangeLog for details)

Non-backwards compatible changes
================================

- output format options specified in filenames: new prototypes for load and
  save tiff and jpeg
- C++ API name changes ... error => VError
- include path change
    - C includes <vips/vips.h>
    - C++ includes <vips/vipscpp.h>
- im_extract*() numbers bands from zero, not from 1 ... also other funcs which
  refer to bands (eg. im_lrmosaic() etc.)
- mosaic funcs have extra max blend width parameter

VIPS enhancements
=================

- rewritten and much fancier ip (see separate docs), now called nip ... old ip
  (slightly fixed) still there
- mmap windows remove 2GB image size limit
- images have an origin hint field
- icc transforms supported via wrapping of Little CMS
- much, much faster Fourier transforms via wrapping of libfwfft
- ppm/pgm/pbm read/write
- C++ API enhancements: new constructors, more operator equivalences
- new colourspace: Yxy
- support for D50 colour temperature images
- new functions: im_image(), im_system(), im_version*(), im_blend(),
  im_copy_set(), im_flood_blob(), im_icc_*(), im_open_local_array(),
  im_header_*(), im_sign(), im_ceil(), im_floor(), im_remainderconst*(),
  im_print(), im_guess_prefix(), im_remosaic(), im_invertlut(), Yxy funcs,
  color temperature funcs, im_clip2fmt(), im_blend(), im_lab_morph(),
  im_histnorm(), im_histcum(), im_video_v4l(), im_LabS2Lab(), im_Lab2LabS()
- new type: IMAGEVEC
- header is much faster
- ip/nip split to separate packages
- better vips2dj
- better cygwin support, mingw support too
- Mac OS X and Windows build support
- new set of sensible #define names, old names can be turned off with
  -DIM_NO_VIPS7_COMPAT
- many configure improvements

Bug fixes
=========

- speed ups
- fixes to implicit image format conversions 
- im_zoom() spots integer overflow
- im_LabS2LabQ() rounding on a/b slightly broken for a/b == 0
- fixes to refcounting in C++ API
- mask casts in C++ were broken




VIPS Changes for 7.4 to 7.6

Non-backwards compatible changes
================================

- im_histplot() has new (and better) rules
- im_andconst(), im_orconst() and im_eorconst() now take a
  double, not an unsigned char, as their constant argument type

libsrc/arithmetic/im_stats.c
/* @(#) im_stats: find general image statistics for all bands separately
(C) Kirk Martinez 1993
23/4/93 J.Cupitt
	- adapted to partial images
	- special struct abandoned; now returns DOUBLEMASK.
1/7/93 JC
	- adapted for partial v2
	- ANSIfied
27/7/93 JC
	- init of mask changed to use actual values, not IM_MAXDOUBLE and
	  (-IM_MAXDOUBLE). These caused problems when assigned to floats.
	  funny business with offset==42, yuk!
31/8/93 JC
	- forgot to init global max/min properly! sorry.
21/6/95 JC
	- still did not init max and min correctly --- now fixed for good

 * 13/1/05
 *	- use 64 bit arithmetic 

*/

/*

    This file is part of VIPS.
    
    VIPS is free software; you can redistribute it and/or modify
    it under the terms of the GNU Lesser General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

 */

/*

    These files are distributed with VIPS - http://www.vips.ecs.soton.ac.uk

 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif /*HAVE_CONFIG_H*/
#include <vips/intl.h>

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <assert.h>

#include <vips/vips.h>

#ifdef WITH_DMALLOC
#include <dmalloc.h>
#endif /*WITH_DMALLOC*/

/* Make and initialise a DOUBLEMASK suitable for grabbing statistics.
 */
static void *
make_mask( IMAGE *im )
{
	DOUBLEMASK *out;

	/* Make temp output.
	 */
	if( !(out = im_create_dmask( "stats", 6, im->Bands + 1 )) )
		return( NULL );
	
	/* Set offset to magic value: 42 indicates we have not yet initialised
	 * max and min for this mask.
	 */
	out->offset = 42;

	return( (void *) out );
}

/* Merge a temp DOUBLEMASK into the real DOUBLEMASK. Row 0 is unused, row 1
 * has the stats for band 1. These are: (minimum, maximum, sum, sum^2). If the
 * offset of out if 42, then it has not been inited yet and we just copy.
 */
static int
merge_mask( DOUBLEMASK *tmp, DOUBLEMASK *out )
{
	double *rowi, *rowo;
	int z;

	/* Merge, or just copy?
	 */
	if( out->offset != 42 )
		/* Add info from tmp.
		 */
		for( z = 1; z < tmp->ysize; z++ ) {
			rowi = tmp->coeff + z * 6;
			rowo = out->coeff + z * 6;

			rowo[0] = IM_MIN( rowi[0], rowo[0] );
			rowo[1] = IM_MAX( rowi[1], rowo[1] );
			rowo[2] += rowi[2];
			rowo[3] += rowi[3];
		}
	else {
		/* Copy info from tmp.
		 */
		for( z = 1; z < tmp->ysize; z++ ) {
			rowi = tmp->coeff + z * 6;
			rowo = out->coeff + z * 6;

			rowo[0] = rowi[0];
			rowo[1] = rowi[1];
			rowo[2] = rowi[2];
			rowo[3] = rowi[3];
		}

		out->offset = 0;
	}

	/* Can now free tmp.
	 */
	im_free_dmask( tmp );

	return( 0 );
}

/* Loop over region, adding information to the appropriate fields of tmp.
 * We set max, min, sum, sum of squares. Our caller fills in the rest.
 */
static int
scan_fn( REGION *reg, DOUBLEMASK *tmp )
{
	Rect *r = &reg->valid;
	IMAGE *im = reg->im;
	int bands = im->Bands;
	int le = r->left;
	int ri = IM_RECT_RIGHT(r);
	int to = r->top;
	int bo = IM_RECT_BOTTOM(r);
	int x, y, z;

/* What type? First define the loop we want to perform for all types.
 * We scan lines bands times to avoid repeating band loops.
 * Use temp variables of same type for min/max for faster comparisons.
 * Use double to sum bands.
 */
#define non_complex_loop(TYPE) \
	{	TYPE *p, *q; \
		TYPE value, small, big; \
		double *row; \
 		\
		/* Have min and max been initialised? \
		 */ \
		if( tmp->offset == 42 ) { \
			/* Init min and max for each band. \
			 */ \
			p = (TYPE *) IM_REGION_ADDR( reg, le, to ); \
			for( z = 1; z < bands + 1; z++ ) { \
				row = tmp->coeff + z * 6; \
				row[0] = p[z - 1]; \
				row[1] = p[z - 1]; \
			} \
			tmp->offset = 0; \
		} \
		\
		for( y = to; y < bo; y++ ) { \
			p = (TYPE *) IM_REGION_ADDR( reg, le, y ); \
 			\
			for( z = 0; z < bands; z++ ) { \
				q = p + z; \
				row = tmp->coeff + (z + 1)*6; \
				small = row[0]; \
				big = row[1]; \
				\
				for( x = le; x < ri; x++ ) { \
					value = *q; \
					q += bands; \
					row[2] += value;\
					row[3] += (double)value*(double)value;\
					if( value > big ) \
						big = value; \
					else if( value < small ) \
						small = value;\
				}\
 				\
				row[0] = small; \
				row[1] = big; \
			}\
		}\
	} 

	/* Now generate code for all types. 
	 */
	switch( im->BandFmt ) {
	case IM_BANDFMT_UCHAR:	non_complex_loop(unsigned char); break; 
	case IM_BANDFMT_CHAR:	non_complex_loop(signed char); break; 
	case IM_BANDFMT_USHORT:	non_complex_loop(unsigned short); break; 
	case IM_BANDFMT_SHORT:	non_complex_loop(signed short); break; 
	case IM_BANDFMT_UINT:	non_complex_loop(unsigned int); break; 
	case IM_BANDFMT_INT:	non_complex_loop(signed int); break; 
	case IM_BANDFMT_DOUBLE:	non_complex_loop(double); break;
	case IM_BANDFMT_FLOAT:	non_complex_loop(float); break; 

	default: 
		assert( 0 );
	}

	return( 0 );
}

/* Find the statistics of an image. Take any non-complex format. Write the
 * stats to a DOUBLEMASK of size 6 by (in->Bands+1). We hold a row for each 
 * band, plus one row for all bands. Row n has 6 elements, which are, in 
 * order, (minimum, maximum, sum, sum^2, mean, deviation) for band n. Row 0 has 
 * the figures for all bands together.
 */
DOUBLEMASK *
im_stats( IMAGE *in )
{	
	DOUBLEMASK *out;
	double *row, *base;
	gint64 pels, vals, z;

	/* Check our args. 
	 */
	if( im_pincheck( in ) )
		return( NULL );
	if( im_iscomplex( in ) ) {
		im_error( "im_stats", _( "bad input type" ) );
		return( NULL );
	}
	if( in->Coding != IM_CODING_NONE ) {
		im_error( "im_stats", _( "not uncoded" ) );
		return( NULL );
	}

	/* Make output.
	 */
	pels = (gint64) in->Xsize * in->Ysize;
	vals = pels * in->Bands;
	if( !(out = make_mask( in )) )
		return( NULL );

	/* Loop over input, calculating min, max, sum, sum^2 for each band
	 * separately.
	 */
	if( im_iterate( in, make_mask, scan_fn, merge_mask, out, NULL ) ) {
		im_free_dmask( out );
		return( NULL );
	}

	/* Calculate mean, deviation, plus overall stats.
	 */
	base = out->coeff;
	base[0] = base[6];	/* Init global max/min */
	base[1] = base[7];
	for( z = 0; z < in->Bands; z++ ) {
		row = base + (z + 1) * 6;
		base[0] = IM_MIN( base[0], row[0] );
		base[1] = IM_MAX( base[1], row[1] );
		base[2] += row[2];
		base[3] += row[3];
		row[4] = row[2] / pels;
		row[5] = sqrt( fabs( row[3] - (row[2] * row[2] / pels) ) / 
			(pels - 1) );
	} 
	base[4] = base[2] / vals;
	base[5] = sqrt( fabs( base[3] - (base[2] * base[2] / vals) ) / 
		(vals - 1) );

	return( out );
}
